### ISSUE TITLE
Translate app UI using AI tools

### ISSUE THREAD

**Issue by benbucksch** [mustang-im/mustang#108]
Once #67 is complete, and we have an English locale generated, we need to translate the app into at least 12 to 20 languages.

### Requirements

Find an online service dedicated to translating UI strings that:
* Can import the locale string format that we use in #67, and export back to the same format. Nice to have (not a requirement): For bonus points, it can directly import from and export to GitHub repos.
* Take the English locale for strings to be translated. Then skips the existing strings that are already translated and mark only the missing strings (which are in the English locale, but not yet in the target locale) as to be translated.
* Offers an online web tool to allow human translators to provide the missing translations for the strings needed. The human translators are per language.
* Offers automatic translations using AI tools (similar to deepl.com and Google Translate) to translate the UI strings. Ideally, the tool would be optimized for translating UI strings. It should offer at least the 20 most common languages.
* Ideally, each string has a state, like "not translated", "draft" and "approved", or similar. The tool should know about software versions. Each software version should show statistics of how many strings are not translated, draft, or approved.
* It would be nice, if the tool was Open-Source. Paid tools are acceptable, as long as the prices are reasonable.

### Plan
1. We will run the translation tool over the strings and generate a first translation for each locale using AI.
2. We will hire human translators that will check and correct the strings that are not translated well. The translator can approve individual strings, or correct them.
3. Export from the tool back into our app repo, and release.

**Comment by jermy-c** [comment]
### Weblate troubleshooting

**Conflict Resolution**

1.  Download Translation Memory for backup
2. Commit changes to weblate repo
3. `git remote add weblate https://hosted.weblate.org/git/project/component/` in the GitHub source code repo
4. `git remote update weblate`
5. `git merge weblate/main`, resolve any conflicts
6. `git rebase origin/main`, resolve any conflicts
7. `git push`
8. Reset repo in the Weblate Repository Maintenance section
9.  Weblate pull

**Can't push to translation branch**

1.  Delete `translation` branch
2.  Push again

**Comment by jermy-c** [comment]
### TODO

- [ ] Have Weblate parse ICU plurals as plurals so it's easier for translators to translate and the AI translation will be more accurate. The docs say it's not supported JSON files. https://docs.weblate.org/en/latest/user/translating.html#plurals https://docs.weblate.org/en/latest/formats.html#translation-types-capabilities

**Comment by benbucksch** [comment]
We cannot always wait for translators for every string addition. We do not have translators for all languages that are constantly standing waiting. We can hire translators only maybe once per year.

Can we please use the following workflow?
1. Extract strings
2. Push to automatic translation
3. Pull automatic translations into translated files (added, in the right sort order).
4. Git commit
5. Push to Weblate
6. Release software
7. Let translators double-check (typically many weeks months later)
8. Pull from weblate to git code repo
9. Release software

Steps 1, 2+3 (together), 5, 8 should be 4 separate commands that we can launch on the command line.
There should be a single GitHub Action for steps 1-4 (all at once).

**Comment by jermy-c** [comment]
After doing some research, it seems like Weblate is prone to merge conflicts when ever the Weblate repo is out of sync with the upstream repo. This would happen even if you don't do a force push.

Why is this a problem? Because to fix it, you would lose some translations in the process:

1. Reset all changes in the Weblate repository: you would lose the translations by the Translators
2. Force writing all translations to the Weblate repository: you would lose the translations by the Translators
3. Rebasing the upstream repository with Weblate changes: you would lose the translations by Auto Translate

And you can't resolve the conflicts:

- Remotely: Weblate doesn't offer a client or API to do that, and 
- Locally: There's no way to push to the Weblate repo after resolving the conflicts locally

**Comment by jermy-c** [comment]
### Progress on workflow so far

> 1. Extract strings

This is finally working well with minimal errors finally with #459 

> 2. Push to automatic translation
> 3. Pull automatic translations into translated files (added, in the right sort order).
> 4. Git commit

Auto translation is working with `json-autotranslate`. I just need to add a feature to include the comments as context.

> 5. Push to Weblate
> 6. Release software
> 7. Let translators double-check (typically many weeks months later)
> 8. Pull from weblate to git code repo
> 9. Release software

Weblate is very prone to conflicts especially when there's many changes to the translation and they happen to be out of sync. I think we could use the either `extractor.ts` code or `json-autotranslate` to push and pull changes from Weblate.

**Comment by benbucksch** [comment]
Yes, manually pushing and pulling from WebLate would be good. Their git connector is really unworkable.

Thanks for the status update!

**Comment by jermy-c** [comment]
## Proposed Solution for Weblate

**Repository**
Let's not push the entire GitHub repo to Weblate instead we just use the `No remote repository` option and push the translations using the Rest API.

### Workflow

**Pull**

1. Rest API [downloads](https://docs.weblate.org/en/latest/api.html#get--api-translations-(string-project)-(string-component)-(string-language)-file-) the translations from Weblate
2. Parse the downloaded file into an Object
3. Replace existing translations with translations from Weblate and add new translations if they don't exists. This avoids too many changes to a file since it's the translations getting pulled not the entire file.

**Review**

1. Review on locally on Git
2. Resolve conflicts that makes sense

**Push**

1. Parse all files
2. Rest API [uploads](https://docs.weblate.org/en/latest/api.html#post--api-translations-(string-project)-(string-component)-(string-language)-file-) them with the param [`string-conflicts: replace-approved`](https://docs.weblate.org/en/latest/user/files.html#conflicts-handling) and [`import-method: replace`](https://docs.weblate.org/en/latest/user/files.html#import-methods) to make sure the reviewed changes are applied

### Pros and Cons

**Pros**

- You have more control over how to resolve conflicts
- Prevents too many changes to a file since the translations are independently applied instead of overwriting the entire file

**Cons**

- Translation strings on Weblate would not always be up-to-date
- Pull Step 3 means that if strings are not reviewed properly or if there's any mistakes on Weblate, it would be overwritten with those changes.

**Comment by benbucksch** [comment]
Yes, manually pushing and pulling from WebLate would be good. Their git connector is really unworkable.

Thanks for the status update!

**Comment by benbucksch** [comment]
The last point of Cons makes this impractical. I also am not sure about WebLate. The UI is not very convenient, either. Maybe there are better tools?

**Comment by jermy-c** [comment]
I just found one today on Reddit: https://www.accent.reviews/

I'll need to see how it works and the features. Otherwise tolgee.io but the self-hosted license restricts to 10 users/seats.

Also how should the tool handle conflicts ideally? Prefer the change stored on the tool or the change on the upstream? Or keep it as is and mark as conflict? <- I think the last solution this is rarely seen.

**Comment by jermy-c** [comment]
## Accent

[Github Repo](https://github.com/mirego/accent)

### Overview

It has a very simple and fast UI similar to Tolgee. There's a great CLI tool to push and pull the translations and it's designed to work well with Github Actions.

### Stack

- Backend: Elixir
- Frontend: Typescript

### Pros

- Simple UI
- Feels fast
- It seems to be dealing with the conflicts well, but if there's a different version on the upstream then it would overwrite with the upstream version
- Features include: Version Control, Collaborators, and commenting
- License does not restrict number of user/seats like Tolgee
- It has a Docker Image. (But I had some difficulties getting it running probably because I'm on ARM, I saw errors stating that)

### Cons

- Lack ICU Message format syntax highlighting and checking
- Lacks other checks placeholder same count

**Comment by benbucksch** [comment]
Accent sounds great. Unfortunately, ICU, placeholders and pluralform are must-haves for us. That was the problem with Weblate - bad translations of pluralform and placeholders destroy our translations and our UI.

**Comment by benbucksch** [comment]
We now have great and well-working AI translations using json-autotranslate and DeepL, so let's close this issue.
